include/boost/beast/http/impl/read.ipp

```diff
@@ -112,72 +126,61 @@ operator()(
     error_code ec,
     std::size_t bytes_transferred)
 {
-    switch(state_)
+    size_t consumed = 0;
+    boost::optional<typename
+        DynamicBuffer::mutable_buffers_type> mb;
+
+    BOOST_ASIO_CORO_REENTER(*this)
     {
-    case 0:
-        state_ = 1;
-        if(b_.size() == 0)
-            goto do_read;
-        goto do_parse;
-
-    case 1:
-        state_ = 2;
-    case 2:
-        if(ec == boost::asio::error::eof)
+        if (b_.size() > 0)
         {
-            BOOST_ASSERT(bytes_transferred == 0);
-            if(p_.got_some())
+            consumed = p_.put(b_.data(), ec);
+            total_bytes_consumed_ += consumed;
+            b_.consume(consumed);
+            if (ec == http::error::need_more)
             {
-                // caller sees EOF on next read
                 ec.assign(0, ec.category());
-                p_.put_eof(ec);
-                if(ec)
-                    goto upcall;
-                BOOST_ASSERT(p_.is_done());
-                goto upcall;
             }
-            ec = error::end_of_stream;
-            goto upcall;
+            BOOST_ASIO_CORO_YIELD
```

Comment:
I think you might need braces here, because `BOOST_ASIO_CORO_YIELD` is a compound statement:
```
if (ec == http::error::need_more)
{
    BOOST_ASIO_CORO_YIELD
    s_.get_io_service().post(
        bind_handler(std::move(*this), ec, 0));
}
```

This is why I indent it the way I do, I treat `BOOST_ASIO_CORO_YIELD` as if it was a separate statement:
https://github.com/boostorg/beast/blob/0bc2a414043f4a9430e71e6664d592e67ea6d282/include/boost/beast/websocket/impl/read.ipp#L162